### ISSUE TITLE
Getting Error: ifcgeom_schema_agnostic/Kernel.h:35:10: error: too few arguments provided to function-like macro invocation #include INCLUDE_SCHEMA(0)

### ISSUE THREAD

**Issue by FlorianB-DE** [IfcOpenShell/IfcOpenShell#4566]
I have said problem: in the file where I have the include, I have:
```cpp
#include <ifcparse/Ifc4.h>
#define IfcSchema Ifc4

#define SCHEMA_VERSIONS = (4)

#include <ifcgeom/IfcGeom.h>
#include <ifcgeom_schema_agnostic/Serialization.h>
```

I also tried as shown in the examples: 
```cpp
#include <string>
#include <iostream>
#include <fstream>

#include <TColgp_Array2OfPnt.hxx>
#include <TColgp_Array1OfPnt.hxx>
#include <TColStd_Array1OfReal.hxx>
#include <TColStd_Array1OfInteger.hxx>

#include <Geom_BSplineSurface.hxx>

#include <BRepBuilderAPI_MakeFace.hxx>
#include <BRepBuilderAPI_NurbsConvert.hxx>

#include <BRepPrimAPI_MakeBox.hxx>
#include <BRepPrimAPI_MakeSphere.hxx>

#include <BRepAlgoAPI_Cut.hxx>

#include <Standard_Version.hxx>

#include "ifcparse/Ifc4.h"
#define IfcSchema Ifc4


#include "ifcparse/IfcBaseClass.h"
#include "ifcparse/IfcHierarchyHelper.h"
#include "ifcgeom/IfcGeom.h"
#include "ifcgeom_schema_agnostic/Serialization.h"

```

to no avail!

**Comment by aothms** [comment]
I've tried a couple of things but couldn't reproduce an error

- Is there a larger code snippet by which I can reproduce this?
- What compiler are you using?
- What are you trying to do? Generally the idea is that as an application developer you don't need to deal with schema specific stuff. Just include only from the ifcgeom_schema_agnostic folder and have it dispatch transparently for you based on the parsed schema at runtime.

**Comment by FlorianB-DE** [comment]
hey @aothms ,
sorry I've lost my cool there and didn't provide any way to reproduce. Sometimes sitting on a problem like 20h is not very helping social skills doesn't it :D 

I'm currently trying to get a bounding box from an IfcSpace. First I've tried getting it only with IfcParse but I think there is no way to to compute that (without making a formular for ecery possibility of IfcRepresentationItem like IfcExtrudedAreaSolid). 

You can look at this project https://github.com/OpenBimRL/OpenBimRL-Engine-Native/tree/working_on_it. Just try to compile it with cmake and you'll get that error :/

**Comment by FlorianB-DE** [comment]
Also boost:
`libboost-all-dev is already the newest version (1.74.0.3).`

**Comment by aothms** [comment]
Can you point me to the exact file in that repo?

If it were just for geometry I would simply try not to have any schema specific code in the program. Use ifcgeom/Iterator to iterate over the geometrical elements. As you say, it does all the evaluation of representation items and opening subtractions and feeds you a bunch vertices. From those vertices you can get the bounding box lower and upper corners. You can look at IfcConvert.cpp and WavefrontOBJSerializer.cpp, but zoom in on the usage of Iterator and TriangulationElement.

We also have some prototypical code btw to fetch property sets in a type safe, but schema agnostic way. https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/examples/IfcParseExamples.cpp#L223-L226 But it relies on the same mechanism of boost preprocessor that you have trouble getting to work.

**Comment by FlorianB-DE** [comment]
the error is thrown here when i include anything from ifcgeom or ifcgeom_schema_agnostic https://github.com/OpenBimRL/OpenBimRL-Engine-Native/blob/working_on_it/include/utils.h

**Comment by aothms** [comment]
Ok, but that is a bit different. You have an `=` sign. And the `#define IfcSchema` uses parenthesis. I think this should be fine:

```C++
#define IfcSchema Ifc4

#define SCHEMA_VERSIONS (2x3)(4)
#define SCHEMA_SEQ (2x3)(4)
```

**Comment by FlorianB-DE** [comment]
Oh well the `=` might actually be the issue here. Ima go test

**Comment by FlorianB-DE** [comment]
Seems to work, thank you very much 🥳! But I do have one question left: Why is there effectively no documentation on the C++ side of things? I mean probably everyone is gonna use python but for me that wasn't really an option. Other than just the exported stuff (https://ifcopenshell.github.io/docs/) which are... not very useful. There is no documentation, is it?

**Comment by FlorianB-DE** [comment]
P.S. Sorry to have bothered you. Hopefully this is useful to anyone running into that mistake again. I saw someone with that issue online but his fix was to downgrade to v0.6.0 and I didn't wanna do that 😅

**Comment by aothms** [comment]
>  Why is there effectively no documentation on the C++ side of things?

Well, you name one reason. But also, the C++ code has also been rather fluent because it underwent heavy changes in v0.6 when we provided support for multiple schemas at runtime and in v0.8 when we provide support for multiple geometry libraries. v0.8 is a good time to begin more extensively documenting things. All help is welcome, very much so.
